### DISCUSSION TITLE
laser logic reversed

### DISCUSSION THREAD

**Discussion by whistlemeup08** [arkypita/LaserGRBL#1215] [Q&A] [answered]
Purchased generic PWM laser that fits 3018 Pro machine to get more power.  Laser is on at power up, on during travel, off during burn.  Using LaserGRBL 1.1.  Set $32 =1 .  No option M4.  Marking on original laser uses Low=on, High=off.

**Comment by arkypita** [comment, marked as answer]
Hi
This is a firmware issue, LaserGRBL cannot change this behaviour.
Google "grbl invert pwm" or "grbl invert spindle" you can find some results.

If I remember you should build a modified version of grbl firmware and flash to your board. In any case, also if you are able to invert the controller board logic to keep the pin HIGH when off and low when ON, there will always be a few milliseconds during board power up, in which grbl has not been loaded, that the pin is low and laser will continue to light up at maximum power.

It would be better if you build a small electronic circuit to reverse the logic of your laser module.

![image](https://user-images.githubusercontent.com/8782035/102658778-877e1480-4178-11eb-8e53-750ea6e481ae.png)

Any NPN transistor is ok (2N3904, BC547, 2N2222...) an resistor value are not critical (but 1K is ok).

PS: It's your module that has the logic reversed, not LaserGRBL nor GRBL!!!

**Reply by arkypita** [reply]
NOTE: VCC means 5V

If you put 12V into VCC you have 12V in PWM_Out. If you want to use 12V in VCC just put a 820ohm resistor in parallel to transistor (pin 3 - pin 1) to have a resistor voltage divider, so you have about 5.4V at PWM_Out

**Reply by whistlemeup08** [reply]
Thanks.  Was hoping for a program change.  I can do this, however.

**Comment by arkypita** [comment]
> Was hoping for a program change. I can do this, however.

Not with a program (lasergrbl) change, but with a firmware change it can be done.

However, for your safety, a logic where 0 mean "ON" is not a safe logic.
It is like a car that accelerates if you release the pedal: doesn't make much sense, what do you think?

**Reply by whistlemeup08** [reply]
Original laser on 3018 Pro marked 1=on, 0= off, which is the reversed general logic.  Thanks for the transistor design.  Not fully working yet as PWM pin on new laser has to go to full ground before it turns off.  Appears not to allow small voltage drop acrossed the transistor.  Will get it working.

**Reply by arkypita** [reply]
> Original laser on 3018 Pro marked 1=on, 0= off, which is the reversed general logic.

Maybe a typo? You mean "marked 0=on, 1= off" (this show the pictures)

> Not fully working yet as PWM pin on new laser has to go to full ground before it turns off

A transistor working in saturation has about 0.2V between collector and emitter.

**Comment by whistlemeup08** [comment]
![20201228_081202](https://user-images.githubusercontent.com/76218139/103216742-bf0d5f00-48e4-11eb-98ba-94b9fa87e6c7.jpg)
![20201228_081238](https://user-images.githubusercontent.com/76218139/103216751-c7fe3080-48e4-11eb-8fc1-41f48373836a.jpg)

**Comment by StuartB4** [comment]
Apart from the laser not working correctly it looks like your Z axis bearings have started to slide up the rods out of the plastic housing. Thats a common fault with the 3018 machines. Have a look on the Sainsmart facebook group, there are several ways to stop it happening.

**Comment by whistlemeup08** [comment]
Diffinite type "O".  I knew better.  Original Logic in my opinion is backwards.  Transistor did not solve problem as laser still on during travel.  Switching laser to different control board and will test.  
Sent from Yahoo Mail on Android 
 
  On Mon, Dec 28, 2020 at 9:27 AM, arkypita<notifications@github.com> wrote:   



Original laser on 3018 Pro marked 1=on, 0= off, which is the reversed general logic.


Maybe a typo? You mean "marked 0=on, 1= off" (this show the pictures)


Not fully working yet as PWM pin on new laser has to go to full ground before it turns off


A transistor working in saturation has about 0.2V between collector and emitter.
